### PR TITLE
Shivansh | Stage | Create a downloads page under other category of all items page

### PR DESCRIPTION
**Pull request by shivanshRawat14**
https://app.clickup.com/t/2hctrbd

### CODE REVIEW

File: apps/web-giddh/src/app/downloads/components/downloads-json/downloads-json.component.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { Component, Inject, OnInit, ViewChild, ElementRef } from "@angular/core";
+import { MatDialogRef, MAT_DIALOG_DATA } from "@angular/material/dialog";
+import * as jsonTreeViewer from 'json-tree-viewer';
+
+@Component({
+    selector: 'downloads-json',
+    templateUrl: './downloads-json.component.html',
+    styleUrls: ['./downloads-json.component.scss']
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads-json/downloads-json.component.ts]: add changeDetectionStrategy

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+     *
+     * @memberof DownloadsComponent
+     */
+    public openDialog(row: any): void {
+        const dialogRef = this.dialog.open(DownloadsJsonComponent, {
+            data: row?.filters,
+            panelClass: 'download-json-panel'
+        });
+
+        dialogRef.afterClosed().subscribe(result => {
```
Comment (the-prateek-mishra) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: Avoid console logs

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.html [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                    <input
+                        type="text"
+                        name="selectedDateRange"
+                        [value]="selectedDateRangeUi ? selectedDateRangeUi : ''"
+                        class="giddh-datepicker date-range-picker"
+                    />
+                </div>
+            </div>
+            <div class="clear-filter form-group mb-0" *ngIf="showDateReport">
+                <button mat-raised-button class="btn-success min-width-100 mr-l1" (click)="resetFilter()">
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.html]: use same class here which is getting used everywhere. currently it's of green color

File: apps/web-giddh/src/app/downloads/components/downloads-json/downloads-json.component.html [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<ng-container
+    appTranslate
+    [file]="'downloads'"
+    (commonLocaleData)="commonLocaleData = $event"
+    (localeData)="localeData = $event"
+>
+    <div class="download-list-container">
+        <div class="modal-header d-flex align-items-center activity-header">
+            <h3 _ngcontent-kab-c237="">{{ localeData.filter }}</h3>
```
Comment (the-prateek-mishra) [inline apps/web-giddh/src/app/downloads/components/downloads-json/downloads-json.component.html]:
```suggestion
            <h3>{{ localeData.filter }}</h3>
```

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.html [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                                element.status === 'INITIATED'
+                            "
+                        >
+                            {{ element.status }}
+                        </div>
+                        <button
+                            *ngIf="element.status === 'COMPLETED'"
+                            mat-button
+                            class="download-button-status btn-success"
+                            (click)="element?.status === 'COMPLETED' ? downloadData(element) : ''"
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.html]: no need of condition here

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            console.log(`Dialog result: ${result}`);
+        });
+    }
+
+    /**
```
Comment (the-prateek-mishra) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts:115]: Do format the code

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    expireAt?: any;
+    requestId?: any;
+    status?: any
+}
+/** Hold information of Download  */
+const ELEMENT_DATA: DownloadData[] = [];
+@Component({
+    selector: 'downloads',
+    templateUrl: './downloads.component.html',
+    styleUrls: ['./downloads.component.scss'],
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts:22]: use changeDetectionStrategy

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+                this.downloadRequest.count = response.body.count;
+            } else {
+                this.dataSource = [];
+                this.downloadRequest.totalItems = 0;
+            }
+            this.changeDetection.detectChanges();
+        });
+    }
+
+    /**
```
Comment (the-prateek-mishra) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts:152]: Do format the JS doc

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+* @memberof DownloadsComponent
+*/
+    public pageChanged(event: any): void {
+        if (this.downloadRequest.page !== event.page) {
+            this.downloadRequest.page = event.page;
+            this.getDownloads();
+        }
+    }
+
+    /**
```
Comment (the-prateek-mishra) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts:165]: Do format the JS doc

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+        if (element) {
+            this.dateFieldPosition = this.generalService.getPosition(element.target);
+        }
+        this.modalRef = this.modalService.show(
+            this.datepickerTemplate,
+            Object.assign({}, { class: 'modal-lg giddh-datepicker-modal', backdrop: false, ignoreBackdropClick: false })
+        );
+    }
+
+    public downloadData(data: any): void {
```
Comment (the-prateek-mishra) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: Avoid empty method

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+     * @memberof DownloadsComponent
+     */
+    public ngOnInit(): void {
+        this.imgPath = isElectron ? 'assets/images/' : AppUrl + APP_FOLDER + 'assets/images/';
+        document.querySelector('body')?.classList?.add('download-page');
+        if (this.generalService.voucherApiVersion === 1) {
+            this.router.navigate(['/pages/home']);
+        }
+        /** Universal date observer */
+        this.universalDate$.subscribe(a => {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: don't use single character variable

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+            }
+        });
+    }
+
+    /**
+     * Opens the Sidebar popup
+     *
+     * @memberof DownloadsComponent
+     */
+    public openDialog(row: any): void {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts:108]: row variable is not included in JS doc of this function

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts (hunk 2 above)
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: what's the use of this ?

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+        dialogRef.afterClosed().subscribe(result => {
+            console.log(`Dialog result: ${result}`);
+        });
+    }
+
+    /**
+  * This function will be called when get the Downloads
+  *
+  * @memberof DownloadsComponent
+  */
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: do code formatting

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+    public getDownloads(resetPage?: boolean) {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: add return type

Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: include parameters in JS doc as well

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+        if (resetPage) {
+            this.downloadRequest.page = 1;
+        }
+        this.isLoading = true;
+        this.downloadsService.getDownloads(this.downloadRequest).pipe(takeUntil(this.destroyed$)).subscribe((response) => {
+            this.isLoading = false;
+            if (response && response.status === 'success') {
+                response.body?.items?.forEach((result: any) => {
+                    result.date = moment(result.date, GIDDH_DATE_FORMAT + " HH:mm:ss").format(GIDDH_DATE_FORMAT);
+                    let today = moment().format('DD-MM-YYYY');
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]:
```suggestion
                    let today = moment().format(GIDDH_DATE_FORMAT);
```

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+                    if (result.expireAt > today) {
+                        result.expireAt = moment(result.expireAt, GIDDH_DATE_FORMAT + " HH:mm:ss").format(GIDDH_DATE_FORMAT);
+                    } else {
+                        result.expireAt = "Expired"
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: use translation variable

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+            this.downloadRequest.page = 1;
+        }
+        this.isLoading = true;
+        this.downloadsService.getDownloads(this.downloadRequest).pipe(takeUntil(this.destroyed$)).subscribe((response) => {
+            this.isLoading = false;
+            if (response && response.status === 'success') {
+                response.body?.items?.forEach((result: any) => {
+                    result.date = moment(result.date, GIDDH_DATE_FORMAT + " HH:mm:ss").format(GIDDH_DATE_FORMAT);
+                    let today = moment().format('DD-MM-YYYY');
+                    if (result.expireAt > today) {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: expiry date should be shown even if today

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+        }
+    }
+
+    /**
+ * Call back function for date/range selection in datepicker
+ *
+ * @param {*} value
+ * @memberof DownloadsComponent
+ */
+    public dateSelectedCallback(value?: any, from?: any): void {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts:173]: include all variables in JS doc

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts (hunk 10 above)
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: remove this function if unused

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+
+    /**
+ * To reset applied filter
+ *
+ * @memberof DownloadsComponent
+ */
+    public resetFilter(): void {
+        this.showDateReport = false;
+        //Reset Date with universal date
+        this.universalDate$.subscribe(a => {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: don't use single character variable

File: apps/web-giddh/src/app/app.module.ts [hunk 20]
```diff
@@ -141,6 +142,7 @@ const DEFAULT_PERFECT_SCROLLBAR_CONFIG: PerfectScrollbarConfigInterface = {
         MatSnackBarModule,
         SnackBarModule,
         MatDialogModule,
+        MatTableModule,
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/app.module.ts]: what's the use of this module here?

File: apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts (hunk 10 above)
Comment (dilpreetsinghofficial) [reply apps/web-giddh/src/app/downloads/components/downloads/downloads.component.ts]: Sir, after this is completed I will work on downloading the export

File: apps/web-giddh/src/assets/locale/all-items/en.json [hunk 21]
```diff
@@ -392,6 +392,11 @@
                     "link": "/pages/search",
                     "icon": "icon-search",
                     "description": "Select group and search for a time span"
+                },
+                {
+                    "label": "Download",
+                    "link": "/pages/download",
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/assets/locale/all-items/en.json]:
```suggestion
                    "link": "/pages/downloads",
```

File: apps/web-giddh/src/assets/locale/all-items/hi.json [hunk 22]
```diff
@@ -388,6 +388,11 @@
                     "link": "/pages/search",
                     "icon": "icon-search",
                     "description": "समूह का चयन करें और समय अवधि खोजें"
+                },
+                {
+                    "label": "डाउनलोड",
+                    "link": "/pages/download",
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/assets/locale/all-items/hi.json]:
```suggestion
                    "link": "/pages/downloads",
```

File: apps/web-giddh/src/assets/locale/all-items/mr.json [hunk 23]
```diff
@@ -388,6 +388,11 @@
                     "link": "/pages/search",
                     "icon": "icon-search",
                     "description": "गट निवडा आणि कालावधीसाठी शोधा"
+                },
+                {
+                    "label": "डाउनलोड करा",
+                    "link": "/pages/download",
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/assets/locale/all-items/mr.json]:
```suggestion
                    "link": "/pages/downloads",
```

File: apps/web-giddh/src/assets/locale/downloads/en.json [hunk 24]
```diff
@@ -0,0 +1,10 @@
+{
+    "requested_date": "Requested Date",
+    "user":"User",
+    "no_logs": "No Record Found",
+    "services": "Services",
+    "filter": "Filter",
+    "download": "Download",
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/assets/locale/downloads/en.json]: this is available in common locale already

File: apps/web-giddh/src/assets/locale/downloads/en.json [hunk 25]
```diff
@@ -0,0 +1,10 @@
...
+    "expiry": "Expiry",
+    "status": "Status"
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/assets/locale/downloads/en.json]: this is available in common locale already

File: apps/web-giddh/src/assets/locale/downloads/en.json [hunk 26]
```diff
@@ -0,0 +1,10 @@
+{
+    "requested_date": "Requested Date",
+    "user":"User",
+    "no_logs": "No Record Found",
+    "services": "Services",
+    "filter": "Filter",
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/assets/locale/downloads/en.json]: this is available in common locale already

File: apps/web-giddh/src/assets/locale/downloads/en.json [hunk 27]
```diff
@@ -0,0 +1,10 @@
+{
+    "requested_date": "Requested Date",
+    "user":"User",
+    "no_logs": "No Record Found",
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/assets/locale/downloads/en.json]: this is available in common locale already

File: apps/web-giddh/src/assets/locale/downloads/en.json [hunk 28]
```diff
@@ -0,0 +1,10 @@
...
+    "services": "Services",
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/assets/locale/downloads/en.json]: this is available in common locale already

**Review by ravinderwalkover — CHANGES_REQUESTED** [review]

**Comment by dilpreetsinghofficial** [comment]
PR review done from side

File: apps/web-giddh/src/app/services/service.module.ts [hunk 29]
```diff
@@ -135,6 +136,8 @@ export class ServiceModule {
                 PurchaseOrderService,
                 BulkVoucherExportService,
                 LocaleService,
+                CustomFieldsService,
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/services/service.module.ts]: remove custom field service

**Review by ravinderwalkover — APPROVED** [review]

**Review by ravinderwalkover — CHANGES_REQUESTED** [review]
test everything

**Review by ravinderwalkover — APPROVED** [review]

**Comment by dilpreetsinghofficial** [comment]
self reviewed

**Review by the-prateek-mishra — APPROVED** [review]

File: apps/web-giddh/src/app/services/service.module.ts [hunk 30]
```diff
@@ -60,9 +60,9 @@ import { PurchaseOrderService } from './purchase-order.service';
 import { BulkVoucherExportService } from './bulkvoucherexport.service';
 import { LocaleService } from './locale.service';
 import { PrinterFormatService } from "./printer.format.service";
-import { CustomFieldsService } from "./custom-fields.service";
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/services/service.module.ts:63]: why is this removed?

File: apps/web-giddh/src/app/shared/shared.module.ts [hunk 31]
```diff
@@ -69,6 +69,8 @@ import { GiddhDateRangepickerModule } from '../theme/giddh-daterangepicker/giddh
 import { DeleteTemplateConfirmationModalModule } from '../invoice/templates/edit-template/modals/confirmation-modal/confirmation.modal.module';
 import { CustomFieldsModule } from './header/components/custom-fields/custom-fields.module';
 import { NgxBootstrapSwitchModule } from 'ngx-bootstrap-switch';
+import { MatSlideToggleModule } from '@angular/material/slide-toggle';
+import { CustomFieldsService } from '../services/custom-fields.service';
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/shared/shared.module.ts]: remove this service from this file

**Review by ravinderwalkover — CHANGES_REQUESTED** [review]

**Review by ravinderwalkover — APPROVED** [review]

**Pull request merged**